### PR TITLE
Job-specific traitor items

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR 
Adds support for job-specific traitor items using a whitelist. The items available are tied to the PDA, so if you steal a T chaplains pda you could purchase T chaplain items.

Added the necronomicon as a test item, but more should be added soon.
**Screenshots**

**Changelog**

:cl:
- tweak: The Necronomicon can now be purchased by Traitor Chaplains in the uplink.

### CODE REVIEW

File: Content.Client/Traitor/Uplink/UplinkMenu.xaml.cs [hunk 1]
```diff
@@ -113,6 +114,21 @@ private void AddListingGui(UplinkListingData listing)
...
+                {
+                    if (listing.JobWhitelist.Contains(job.ID))
+                    {
+                        found = true;
+                        break;
+                    }
+                }
+                if (!found)
+                    return;
+            }
```
Comment (metalgearsloth) [inline Content.Client/Traitor/Uplink/UplinkMenu.xaml.cs:130]: You shouldn't send these in the first place.

File: Content.Server/Traitor/Uplink/UplinkSystem.cs [hunk 2]
```diff
@@ -166,9 +169,31 @@ private void UpdateUserInterface(UplinkComponent component)
...
+                    foreach (var role in mind.Mind.AllRoles.ToList())
+                    {
+                        if (role.GetType() == typeof(Job))
+                        {
+                            var job = (Job) role;
+                            jobList.Add(job.Prototype);
+                        }
+                    }
+                    component.JobWhitelist = jobList;
+                }
```
Comment (metalgearsloth) [inline Content.Server/Traitor/Uplink/UplinkSystem.cs:191]: the roles shitcode continues :despair:

File: Content.Server/Traitor/Uplink/UplinkComponent.cs [hunk 3]
```diff
@@ -22,6 +23,8 @@ public sealed class UplinkComponent : Component
 
         [ViewVariables] public UplinkAccount? UplinkAccount;
 
+        [ViewVariables] public HashSet<JobPrototype>? JobWhitelist = null;
```
Comment (metalgearsloth) [inline Content.Server/Traitor/Uplink/UplinkComponent.cs:26]: Just use a string + typeserializer for it then you don't need jobprototype to be serializable. Also this doesn't have a datafield.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Traitor/Uplink/UplinkSystem.cs (hunk 2 above)
Comment (EmoGarbage404) [reply Content.Server/Traitor/Uplink/UplinkSystem.cs:191]: you know it, baby

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**